File: website-backend/src/trackers/ranker.rs
use super::Tracker;
use crate::http::{get_json, HttpClient, JsonValue};
use failure::Error;
use hyper::Uri;
use tokio::prelude::*;

const RANKER_API_URL: &str = "http://api.ranker.com/lists/298553/items/85372114?include=crowdRankedStats,votes";

#[derive(Debug, Eq, PartialEq, serde::Serialize, serde::Deserialize)]
pub struct DataPoint {
  pub rank: u64,
  pub upvotes: u64,
  pub downvotes: u64,
  pub reranks: u64,
  pub top5_reranks: u64,
}

pub struct RankerTracker {
  url: Uri,
}

impl RankerTracker {
  pub fn new() -> Self {
    Self { url: Uri::from_static(RANKER_API_URL) }
  }
}

impl Tracker for RankerTracker {
  type DataPoint = DataPoint;

  fn describe(&self) -> String {
    "ranker".to_owned()
  }

  fn fetch_data_point(
    &self,
    http_client: &HttpClient,
  ) -> Box<dyn Future<Item = Self::DataPoint, Error = Error> + Send> {
    Box::new(get_json(&http_client, self.url.clone()).and_then(
      |json: JsonValue| {
        json_to_data_point(json)
          .ok_or_else(|| failure::err_msg("malformed JSON response from API"))
      },
    ))
  }
}

fn json_to_data_point(json: JsonValue) -> Option<DataPoint> {
  Some(DataPoint {
    rank: json["rank"].as_u64()?,
    upvotes: json["votes"]["upVotes"].as_u64()?,
    downvotes: json["votes"]["downVotes"].as_u64()?,
    reranks: json["crowdRankedStats"]["totalContributingListCount"].as_u64()?,
    top5_reranks: json["crowdRankedStats"]["top5ListCount"].as_u64()?,
  })
}

File: website-backend/src/trackers/ranker.js
const fetch = require('node-fetch');
const typeCheck = require('../utils/typeCheck');

module.exports = {
  name: 'ranker',
  createFetcher: ({ listId, itemId }) => {
    typeCheck.assert(listId, 'listId', 'String');
    typeCheck.assert(itemId, 'itemId', 'String');

    let apiUrl = `https://api.ranker.com/lists/${listId}/items/${itemId}?include=crowdRankedStats,votes`;
    return () =>
      fetch(apiUrl)
        .then(response => response.json())
        .then(data => {
          typeCheck.assert(data, 'data', 'Object');
          let { rank, votes, crowdRankedStats } = data;
          typeCheck.assert(rank, 'rank', 'Number');

          typeCheck.assert(votes, 'votes', 'Object');
          let { upVotes, downVotes } = votes;
          typeCheck.assert(upVotes, 'upVotes', 'Number');
          typeCheck.assert(downVotes, 'downVotes', 'Number');

          typeCheck.assert(crowdRankedStats, 'crowdRankedStats', 'Object');
          let { totalContributingListCount, top5ListCount } = crowdRankedStats;
          typeCheck.assert(
            totalContributingListCount,
            'totalContributingListCount',
            'Number',
          );
          typeCheck.assert(top5ListCount, 'top5ListCount', 'Number');

          return {
            rank,
            upvotes: upVotes,
            downvotes: downVotes,
            reranks: totalContributingListCount,
            top5Reranks: top5ListCount,
          };
        });
  },
};
